rfcs/0062-content-addressed-paths.md

```diff
@@ -79,54 +76,73 @@ The gist of the design is that:
 
 ### Output mappings
 
-A major consequence of allowing content-addressed derivations is that the
-actual output path of a derivation might not match its output hash anymore.
-
-To express this, we introduce a new mapping `pathOf` that associates the hash
-of every live derivation to its store path.
-By extension, we also define `pathOf(drv) = pathOf(hash(drv))`
-
-### Building a ca derivation
-
-ca derivations are derivations with the `__contentAddressed` argument set to
-`true`.
+For each output `output` of a derivation `drv`, we define
 
-The process for building a content-adressed derivation is the following:
+- its output id **DrvOutputId(drv, output)** as the tuple `(hash(drv), output, truster)`, where `truster` is a reserved field for future use and currently always set to `"world"`.
+  This id uniquely identifies the output.
+  We textually represent this as `hash(drv)!output[@truster]`.
+- its concrete path **PathOf(outputId)** as the path on which the output will be stored on disk.
 
-- We build it like a normal derivation (see below) to get an output path `$out`.
-- We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
-- We move `$out` to `/nix/store/$chash-$name`
-- We create a mapping from `$dhash` (the hash computed at eval-time) to
-  `/nix/store/$chash-$name`
+> Unresolved: should we already include the `truster` field in `DrvOutputId`
+> even if it's not used atm? What would be the cost of adding it later?
 
-[^modulo-hashing]:
+In a dependency-addressed-only world, the concrete path for a derivation output was a pure function of this output's id that could be computed at eval-time. However this won't be the case anymore once we allow content-addressed derivations, so we now need to store the results the `PathOf` function in the Nix database as a new table:
 
-  We can possibly normalize all the self-references before
-  computing the hash and rewrite them when moving the path to handle paths with
-  self-references, but this isn't strictly required for a first iteration
+```sql
+create table if not exists PathOf (
+    drv integer not null,
+    output text not null,
+    truster integer not null,
+    path integer not null,
+)
+```
 
 ### Building a normal derivation
 
 #### Resolved derivations
 
-We define a `resolved derivation` as a derivation that has no reference to any
-other derivation (but can refere to store paths).
+We define a **resolved derivation** as a derivation whose only references are either:
 
-For a derivation `drv` whose input derivations have all been realised, we define
-its `associated resolved derivation` of `drv` (`resolved(drv)`) as
-`drv` in which we replace every input derivation `inDrv` of `drv` by
-`pathOf(inDrv)` (and update the output hash accordingly).
+- Self references
```

Comment:
Strictly speaking its the outputs, not the DRVs themselves, that have the self references. DRV self-references are handled with the `placeholder` thing which is different.

Reply:
Yeah, the definition of a self-reference is missing here.